In [1]:
pip list

Package                            Version          Note: you may need to restart the kernel to use updated packages.
---------------------------------- -----------------
absl-py                            0.9.0            
alabaster                          0.7.12           
anaconda-client                    1.7.2            
anaconda-navigator                 1.9.12           
anaconda-project                   0.8.3            
argh                               0.26.2           
asn1crypto                         1.3.0            
astor                              0.8.1            
astroid                            2.3.3            
astropy                            4.0.1.post1      
atomicwrites                       1.3.0            
attrs                              19.3.0           
autopep8                           1.4.4            
Babel                              2.8.0            
backcall                           0.1.0            
backports.functools-lru-cache     

## 개발 환경(OS) 및 라이브러리 버전 
- 운영체제(os) Windows10
- pandas                             0.23.4           
- matplotlib                         3.1.3   
- numpy                              1.18.1   

In [138]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.impute import KNNImputer
import numpy as np

In [176]:
df_cus = pd.read_csv("/data/cus_info.csv")

In [28]:
df_cus.head()

act_id  sex_dit_cd  \
0  64aae8dd71e5c0761000db9f9a6779e504e7a4aa9dc097...           1   
1  5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...           1   
2  1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...           2   
3  6d497facfa1ea5901b827335553331f8555fec02a8184f...           2   
4  b727c78d2cfc246c97b677f29a034399a0c7e7873fff44...           1   

   cus_age_stn_cd  ivs_icn_cd  cus_aet_stn_cd  mrz_pdt_tp_sgm_cd  lsg_sgm_cd  \
0               4          99               1                  2           3   
1               6           4               4                  2           5   
2               7           4               3                  2           5   
3               6           4               4                  8           5   
4               5           2               2                  2           5   

   tco_cus_grd_cd  tot_ivs_te_sgm_cd  mrz_btp_dit_cd  
0               3                  6              16  
1               2                  6               1  
2               5                  6               9  
3               3                  6              16  
4               5                  5              16

In [178]:
df_cus = df_cus.rename({'act_id':'계좌ID', 'sex_dit_cd':'성별', 'cus_age_stn_cd':'연령대',
               'ivs_icn_cd' : '투자성향', 'cus_aet_stn_cd': '자산구간', 
               'mrz_pdt_tp_sgm_cd': '주거래상품군','lsg_sgm_cd':'lifestage',
               'tco_cus_grd_cd':'서비스등급','tot_ivs_te_sgm_cd': '총투자기간',
              'mrz_btp_dit_cd': '주거래업종구분'}, axis = 1)

In [179]:
df_cus.head()

계좌ID  성별  연령대  투자성향  자산구간  \
0  64aae8dd71e5c0761000db9f9a6779e504e7a4aa9dc097...   1    4    99     1   
1  5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...   1    6     4     4   
2  1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...   2    7     4     3   
3  6d497facfa1ea5901b827335553331f8555fec02a8184f...   2    6     4     4   
4  b727c78d2cfc246c97b677f29a034399a0c7e7873fff44...   1    5     2     2   

   주거래상품군  lifestage  서비스등급  총투자기간  주거래업종구분  
0       2          3      3      6       16  
1       2          5      2      6        1  
2       2          5      5      6        9  
3       8          5      3      6       16  
4       2          5      5      5       16

### 고객 데이터 결측값

In [31]:
df_cus['투자성향'].unique()

array([99,  4,  2,  3,  5,  1,  9,  0], dtype=int64)

In [32]:
print('투자성향 결측치 : ', end= '')
print(len(df_cus[df_cus['투자성향'] == 99]) / len(df_cus)*100)
print('서비스등급 결측치 : ', end= '')
print(len(df_cus[df_cus['서비스등급'] == 99]) / len(df_cus)*100)

투자성향 결측치 : 36.870000000000005
서비스등급 결측치 : 0.37


In [33]:
len(df_cus[df_cus['투자성향'] == 99].index)

3687

In [34]:
len(df_cus[df_cus['서비스등급'] == 99].index)

37

In [35]:
len(df_cus)

10000

### 주식 종복에 대한 코드 정보 결측값
- 시가총액 규모유형
- 시장구분

In [16]:
df_iem = pd.read_csv("/data/iem_info.csv")

In [17]:
df_iem = df_iem.rename({'iem_cd': '종목코드', 'iem_krl_nm': '종목한글명','btp_cfc_cd' : '종목업종',
              'mkt_pr_tal_scl_tp_cd': '시가총액규모유형','stk_dit_cd':'시장구분'}, axis=1)

In [18]:
df_iem.head(3)

종목코드   종목한글명  종목업종  시가총액규모유형  시장구분
0  A000020    동화약품     8         2    99
1  A000080   하이트진로    14         2     1
2  A000180  성창기업지주     5         3    99

In [19]:
df_iem.shape

(3079, 5)

In [25]:
print('시가총액규모유형 : ', end = '')
print(len(df_iem[df_iem['시가총액규모유형'] == 99])/len(df_iem)*100, end = '')
print('%')
print('시장구분 : ', end = '')
print(len(df_iem[df_iem['시장구분'] == 99])/len(df_iem)*100, end = '')
print('%')

시가총액규모유형 : 25.62520298798311%
시장구분 : 88.60019486846379%


### 종합
- 투자성향 : 대체 
- 서비스등급 : 결측치 처리 안함 
- 시가총액규모유형 : 대체 
- 시장구분 : drop 

In [36]:
print('투자성향 결측치 : ' + str(len(df_cus[df_cus['투자성향'] == 99]) / len(df_cus)*100) + '%')
print('서비스등급 결측치 : ' + str(len(df_cus[df_cus['서비스등급'] == 99]) / len(df_cus)*100) + '%')

투자성향 결측치 : 36.870000000000005%
서비스등급 결측치 : 0.37%


In [37]:
print('시가총액규모유형 : ' + str(len(df_iem[df_iem['시가총액규모유형'] == 99])/len(df_iem)*100)+ '%')
print('시장구분 : ' + str(len(df_iem[df_iem['시장구분'] == 99])/len(df_iem)*100)+ '%')

시가총액규모유형 : 25.62520298798311%
시장구분 : 88.60019486846379%


### 결측값 처리 
- 행을 지우는 건 지양 : 계좌id가 없어질 수 있음 
- 시가총액규모유형 : 유가증권시장의 경우 6월부터 8월 마지막 영업일까지 하루평균 시가총액 순서로 1~100위는 대형주, 101~300위는 중형주, 그 외는 소형주로 분류. 코스닥시장은 중형주 기준이 101~400위. / 유가증권시장(코스피), 코스닥
- 코스피에서 시가총액 순위를 기준으로 만든 대형주, 중형주, 소형주 지수가 있음, 대형주 시가총액 1~100위, 시가총액 101 ~ 300위, 시가총액 301 ~ 

In [40]:
# 시장구분 열 drop 
df_iem = df_iem.drop('시장구분', axis = 1)

In [41]:
# 투자성향 분포 확인 
df_cus['투자성향'].value_counts()

99    3687
4     1858
3     1669
2     1467
5      935
1      210
0      171
9        3
Name: 투자성향, dtype: int64

#### 시가총액규모유형 전처리 방법 
- 행을 지우는 건 지양 : 주식중목코드가 없어질 수 있음
- KODEX|TIGER|KINDEX|KTOP|KBSTAR|ARIRANG|HANARO|KOSEF 은 파생상품 혹은 펀드이므로, 새로운 라벨 "4" 추가. 
- ~우 인 데이터도 있음.. 05. 우량주 추가 
- 01: 대형주, 02: 중형주, 03: 소형주, 04: 파생상품, 05. 우량주, 99: 기타 

In [43]:
# 시가총액규모유형 분포 확인 
df_iem['시가총액규모유형'].value_counts()

3     1577
99     789
2      512
1      201
Name: 시가총액규모유형, dtype: int64

In [45]:
df_iem_null = df_iem[df_iem['시가총액규모유형'] == 99]

In [46]:
df_iem_null.to_csv("시가총액규모유형_결측치.csv", index = False)

In [73]:
 # https://zephyrus1111.tistory.com/46
df_iem.query('종목한글명.str.contains("KODEX|TIGER|KINDEX|KTOP|KBSTAR|ARIRANG|HANARO|KOSEF")', engine='python')['시가총액규모유형'].replace(99,4, inplace = True)

In [89]:
 # https://zephyrus1111.tistory.com/46
index_04 = df_iem.query('종목한글명.str.contains("KODEX|TIGER|KINDEX|KTOP|KBSTAR|ARIRANG|HANARO|KOSEF")', engine='python')['시가총액규모유형'].index

In [96]:
# 04 파생상품 추가 
for i in index_04: 
    df_iem.iloc[i, -1] = 4
    #print(df_iem.iloc[i, -1])

In [100]:
df_iem['시가총액규모유형'].value_counts()

3     1577
2      512
4      395
99     394
1      201
Name: 시가총액규모유형, dtype: int64

In [106]:
index_05 = []
for i in range(len(df_iem)):
    if df_iem.loc[i, '종목한글명'].endswith('우'):
        index_05.append(i)

In [108]:
for i in index_05:
    df_iem.iloc[i, -1] = 5
    #print(df_iem.loc[i, '종목한글명'])

In [114]:
df_iem['시가총액규모유형'].value_counts()

3     1577
2      511
4      395
99     275
1      200
5      121
Name: 시가총액규모유형, dtype: int64

In [115]:
df_iem.to_csv("iem_info_missingvalue_out.csv", index = False)

In [116]:
df_iem.shape

(3079, 4)

In [117]:
df_iem[df_iem['시가총액규모유형'] == 4]

종목코드                     종목한글명  종목업종  시가총액규모유형
384   A091160                 KODEX 반도체    14         4
385   A091180                 KODEX 자동차    14         4
405   A099140             KODEX China H    14         4
410   A101280          KODEX 일본TOPIX100    14         4
421   A104520                 KOSEF 블루칩    14         4
441   A117700                  KODEX 건설    14         4
445   A122630                KODEX 레버리지    14         4
450   A123310                 TIGER 인버스    14         4
464   A131890           KINDEX 삼성그룹동일가중    14         4
465   A133690            TIGER 미국나스닥100    14         4
473   A139220              TIGER 200 건설    14         4
474   A139250           TIGER 200 에너지화학    14         4
475   A139270              TIGER 200 금융    14         4
476   A139280                TIGER 경기방어    14         4
481   A140710                  KODEX 운송    14         4
486   A143850       TIGER 미국S&P500선물(H)    14         4
493   A152100               ARIRANG 200    14         4
494   A152280              KOSEF 200 선물    14         4
500   A157500                  TIGER 증권    14         4
505   A168300                KTOP 코스피50    14         4
507   A169950            KODEX 중국본토 A50    14         4
512   A174360      KBSTAR 중국본토대형주CSI100    14         4
531   A195930       TIGER 유로스탁스50(합성 H)    14         4
532   A196030     KINDEX 일본TOPIX레버리지(H)    14         4
535   A200030        KODEX 미국S&P산업재(합성)    14         4
536   A200050          KODEX 독일MSCI(합성)    14         4
546   A211560                TIGER 배당성장    14         4
548   A213630  ARIRANG 미국다우존스고배당주(합성 H)    14         4
562   A219480       KODEX 미국S&P500선물(H)    14         4
567   A222180       ARIRANG 스마트베타 Value    14         4
...       ...                       ...   ...       ...
3009  A322150            KINDEX 스마트하이베타    14         4
3012  A325020                KODEX 배당가치    14         4
3014  A326230             KBSTAR 내수주플러스    14         4
3015  A329200           TIGER 부동산인프라고배당    14         4
3016  A329650             KODEX TRF3070    14         4
3017  A332610         ARIRANG 미국단기우량회사채    14         4
3022  A333960         ARIRANG KS모멘텀가중TR    14         4
3023  A334690           KBSTAR 팔라듐선물(H)    14         4
3024  A334700        KBSTAR 팔라듐선물인버스(H)    14         4
3027  A337140              KODEX 코스피대형주    14         4
3032  A341850       TIGER KIS부동산인프라채권TR    14         4
3041  A352560         KODEX 다우존스미국리츠(H)    14         4
3049  A364960       TIGER KRX BBIG K-뉴딜    14         4
3050  A364970          TIGER KRX바이오K-뉴딜    14         4
3055  A367740             HANARO Fn5G산업    14         4
3056  A367770           KBSTAR Fn수소경제테마    14         4
3057  A368190      HANARO Fn K-뉴딜디지털플러스    14         4
3058  A368200      KBSTAR Fn K-뉴딜디지털플러스    14         4
3060  A371150            KBSTAR 차이나항셍테크    14         4
3063  A373490          KODEX K-이노베이션액티브    14         4
3065  A376250          ARIRANG 탄소효율그린뉴딜    14         4
3066  A376410            TIGER 탄소효율그린뉴딜    14         4
3067  A377990            TIGER Fn신재생에너지    14         4
3068  A379790         KBSTAR 유로스탁스50(H)    14         4
3070  A380340            KINDEX Fn5G플러스    14         4
3072  A381560           HANARO Fn전기&수소차    14         4
3075  A385520            KODEX K-미래차액티브    14         4
3076  A388280            KBSTAR Fn컨택트대표    14         4
3077  A388420         KBSTAR 비메모리반도체액티브    14         4
3078  A394350     KOSEF 릭소글로벌퓨처모빌리티MSCI    14         4

[395 rows x 4 columns]

In [118]:
df_iem[df_iem['시가총액규모유형'] == 5]

종목코드      종목한글명  종목업종  시가총액규모유형
4     A000325     노루홀딩스우     2         5
17    A001465       BYC우    14         5
19    A001525        동양우    11         5
42    A003075    코오롱글로벌우     1         5
46    A003535    한화투자증권우     2         5
57    A004255       NPC우    12         5
58    A004365        세방우     6         5
63    A004545     깨끗한나라우    10         5
70    A004985      성신양회우    11         5
73    A005257   녹십자홀딩스2우     5         5
74    A005305      롯데칠성우    14         5
116   A008775      호텔신라우     7         5
121   A009155      삼성전기우     9         5
136   A011155    CJ씨푸드1우    14         5
156   A014915      성문전자우     9         5
163   A016385    KG동부제철우    11         5
175   A019175      신풍제약우     8         5
206   A032685      소프트센우     7         5
426   A108675    LX하우시스우    12         5
489   A145995       삼양사우    14         5
672   A29923L  세븐브로이양평2우    14         5
675   A30018L     팝스라인2우    14         5
691   A31134K   테디아일랜드1우    14         5
714   A32316K       테바1우    14         5
730   A34385K  금축인터내셔널1우    14         5
740   A35320K     대덕전자1우     9         5
761   A38380K    LX홀딩스1우    14         5
766   A000105      유한양행우     8         5
768   A000145  하이트진로홀딩스우     2         5
777   A000545      흥국화재우     2         5
...       ...        ...   ...       ...
2317  A000215        DL우     1         5
2318  A000225     유유제약1우     8         5
2326  A000725      현대건설우     1         5
2339  A001725      신영증권우     2         5
2340  A001795      대한제당우    14         5
2356  A002995      금호건설우     1         5
2365  A003415      쌍용양회우    14         5
2367  A003465      유화증권우     2         5
2370  A003555        LG우     2         5
2378  A004135     대덕GDS우    14         5
2391  A00499K      롯데지주우     2         5
2398  A005745  크라운해태홀딩스우     2         5
2399  A005935      삼성전자우     9         5
2402  A006125   SK디스커버리우     2         5
2408  A006805    미래에셋증권우     2         5
2416  A007575      일양약품우     8         5
2420  A00806K       대덕1우     2         5
2427  A009275        신원우    14         5
2461  A014285      금강공업우    11         5
2534  A03473K        SK우     2         5
2666  A071055    한국금융지주우     2         5
2692  A084695     대상홀딩스우     2         5
2699  A090435    아모레퍼시픽우    12         5
2719  A096775   SK이노베이션우    12         5
2722  A097955   CJ제일제당 우    14         5
2949  A27863K   스터디플래닛1우    14         5
2972  A294090      이오플로우    10         5
3026  A33637K  솔루스첨단소재1우     9         5
3048  A36328K    티와이홀딩스우     2         5
3064  A37550K     DL이앤씨우    14         5

[121 rows x 4 columns]

#### 투자성향 전처리 방법
- knn 으로 99에 해당하는 값 3687개 중에서 1000개만 99로 남김 (99에 대한 학습도 필요하니깐)
- 나머지 값은 null값으로 만듬
- knn 으로 n=3일때의 예측값 대치
- 35, 36 등과 같이 범주형 변수가 아닌 경우, 99로 바꿈

In [216]:
df_cus = pd.read_csv("/data/cus_info.csv")

In [243]:
df_cus.head()

계좌ID  성별  연령대 투자성향  자산구간  \
0  64aae8dd71e5c0761000db9f9a6779e504e7a4aa9dc097...   1    4   99     1   
1  5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...   1    6    4     4   
2  1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...   2    7    4     3   
3  6d497facfa1ea5901b827335553331f8555fec02a8184f...   2    6    4     4   
4  b727c78d2cfc246c97b677f29a034399a0c7e7873fff44...   1    5    2     2   

   주거래상품군  lifestage  서비스등급  총투자기간  주거래업종구분  
0       2          3      3      6       16  
1       2          5      2      6        1  
2       2          5      5      6        9  
3       8          5      3      6       16  
4       2          5      5      5       16

In [218]:
df_cus = df_cus.rename({'act_id':'계좌ID', 'sex_dit_cd':'성별', 'cus_age_stn_cd':'연령대',
               'ivs_icn_cd' : '투자성향', 'cus_aet_stn_cd': '자산구간', 
               'mrz_pdt_tp_sgm_cd': '주거래상품군','lsg_sgm_cd':'lifestage',
               'tco_cus_grd_cd':'서비스등급','tot_ivs_te_sgm_cd': '총투자기간',
              'mrz_btp_dit_cd': '주거래업종구분'}, axis = 1)

In [227]:
df_cus['투자성향'].replace({99, '99'}, inplace = True)

In [238]:
df_cus['투자성향'] = df_cus['투자성향'].astype('str')

In [242]:
type(df_cus['투자성향'][3])

str

In [310]:
df_cus_null = df_cus.copy()

In [311]:
type(df_cus_null['투자성향'][3])

str

In [312]:
index_null = df_cus_null[df_cus_null['투자성향'] == '99'][:-1000].index

In [313]:
len(index_null)

2687

In [314]:
for i in index_null:
    df_cus_null.iloc[i, 3] = np.nan

In [315]:
df_cus_null.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 10 columns):
계좌ID         10000 non-null object
성별           10000 non-null int64
연령대          10000 non-null int64
투자성향         7313 non-null object
자산구간         10000 non-null int64
주거래상품군       10000 non-null int64
lifestage    10000 non-null int64
서비스등급        10000 non-null int64
총투자기간        10000 non-null int64
주거래업종구분      10000 non-null int64
dtypes: int64(8), object(2)
memory usage: 781.3+ KB


In [316]:
# https://www.analyticsvidhya.com/blog/2020/07/knnimputer-a-robust-way-to-impute-missing-values-using-scikit-learn/
imputer = KNNImputer(n_neighbors = 3)

In [317]:
imputed = imputer.fit_transform(df_cus_null.iloc[:, 1:])

In [318]:
df_imputer = pd.DataFrame(imputed, columns = df_cus_null.columns[1:])

In [319]:
df_imputer['투자성향'].value_counts()

4.000000     2104
3.000000     2018
2.000000     1586
99.000000    1021
5.000000      950
3.333333      347
3.666667      296
2.666667      240
1.000000      219
2.333333      205
0.000000      178
4.333333      104
35.000000     101
35.333333      96
34.666667      74
34.333333      64
1.666667       61
35.666667      59
4.666667       46
1.333333       42
66.666667      33
67.000000      30
67.333333      29
34.000000      24
36.000000      23
0.666667       12
33.666667      10
67.666667       9
36.333333       7
66.333333       5
66.000000       3
9.000000        3
5.333333        1
Name: 투자성향, dtype: int64

In [320]:
df_imputer['투자성향'] = df_imputer['투자성향'].astype('int')
df_imputer['성별'] = df_imputer['성별'].astype('int')
df_imputer['연령대'] = df_imputer['연령대'].astype('int')
df_imputer['자산구간'] = df_imputer['자산구간'].astype('int')
df_imputer['주거래상품군'] = df_imputer['주거래상품군'].astype('int')
df_imputer['lifestage'] = df_imputer['lifestage'].astype('int')
df_imputer['서비스등급'] = df_imputer['서비스등급'].astype('int')
df_imputer['총투자기간'] = df_imputer['총투자기간'].astype('int')
df_imputer['주거래업종구분'] = df_imputer['주거래업종구분'].astype('int')

In [321]:
df_imputer['투자성향'].value_counts()

3     2661
4     2254
2     2031
99    1021
5      951
1      322
35     256
0      190
34     162
67      68
66      41
36      30
33      10
9        3
Name: 투자성향, dtype: int64

In [322]:
df_imputer['투자성향'].replace(35, 99, inplace=True)
df_imputer['투자성향'].replace(34, 99, inplace=True)
df_imputer['투자성향'].replace(67, 99, inplace=True)
df_imputer['투자성향'].replace(36, 99, inplace=True)
df_imputer['투자성향'].replace(33, 99, inplace=True)
df_imputer['투자성향'].replace(66, 99, inplace=True)

In [323]:
df_imputer['투자성향'].value_counts()

3     2661
4     2254
2     2031
99    1588
5      951
1      322
0      190
9        3
Name: 투자성향, dtype: int64

In [324]:
df_imputer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 9 columns):
성별           10000 non-null int32
연령대          10000 non-null int32
투자성향         10000 non-null int32
자산구간         10000 non-null int32
주거래상품군       10000 non-null int32
lifestage    10000 non-null int32
서비스등급        10000 non-null int32
총투자기간        10000 non-null int32
주거래업종구분      10000 non-null int32
dtypes: int32(9)
memory usage: 351.6 KB


In [325]:
df_imputer_ver2 = pd.concat([df_cus_null['계좌ID'], df_imputer], axis = 1)

In [327]:
df_imputer_ver2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 10 columns):
계좌ID         10000 non-null object
성별           10000 non-null int32
연령대          10000 non-null int32
투자성향         10000 non-null int32
자산구간         10000 non-null int32
주거래상품군       10000 non-null int32
lifestage    10000 non-null int32
서비스등급        10000 non-null int32
총투자기간        10000 non-null int32
주거래업종구분      10000 non-null int32
dtypes: int32(9), object(1)
memory usage: 429.8+ KB


In [329]:
df_imputer_ver2

계좌ID  성별  연령대  투자성향  자산구간  \
0     64aae8dd71e5c0761000db9f9a6779e504e7a4aa9dc097...   1    4     4     1   
1     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...   1    6     4     4   
2     1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...   2    7     4     3   
3     6d497facfa1ea5901b827335553331f8555fec02a8184f...   2    6     4     4   
4     b727c78d2cfc246c97b677f29a034399a0c7e7873fff44...   1    5     2     2   
5     4f330270cd37d758fbe9dfe45e0c40ed4579ab6b87a648...   1    5     3     4   
6     407a1110ee9394a0d7778e13de9327117182a68e545372...   1    8    99     2   
7     2c17850345fbac9b3f004773972fd72a716228b3896e6f...   2    4     3     4   
8     812f478747ca7497e60a6643ca82a6d5c7323980291a42...   1    4     3     4   
9     51c0d868a0d91c15a5eefdd68fbca1d1c5e3215fcebb33...   1    5     5     3   
10    127188207e133cd6361a6966376eea6183b31077a82bd3...   1    4     3     3   
11    f6bfe08cc25c9583f61e0cb573a6f7a2593e8314732cb9...   1    8     4     6   
12    634c1db2c131ef2f3a404fb533fcda64fa517b0fbe9316...   1    7     3     2   
13    249201e1820bafc4e0b1eaa4d255588c7c109ba390611c...   1    5     4     1   
14    dbc7f1c8694e7a40827680f9d8b8538db774a134fa24bb...   1    7     3     3   
15    5b6973bff9ab839507cf2f839b8b6d190d8ac2144fa5a4...   1    6     5     5   
16    9190b0953de458aea95e2d38b623ba3c9e5785b3695d8c...   1    6     4     2   
17    f2ef679da3d31ff7aa361f753793150a51e82049a55941...   2    6     3     1   
18    141aa640948a63ffa54747efda8272da47e2d8846ee02d...   2    9     2     3   
19    9ec37f3b747828f340de272ee3e06e3fee4a2cf9eba26a...   2    4    99     3   
20    26dd1e72772128f6639abfa9056f39941b33949075c27c...   1    9     3     4   
21    7a17368be2d52cf0c0bd1873748d09ebf416c437835ea1...   1    7     5     6   
22    90ad0b4616a16f5e12d30edf79bec17993a9945dfe43ba...   1    6     3     4   
23    1c52768eebea2eab6a7ed8cc95d838a3732fc6dce51ba3...   2    9     3     5   
24    354e72d40ddf6245425489b47222dc931f1b97350c2b41...   2    5     4     2   
25    39320d0a5c3dc1a965cce964ac10491a731792fd26933e...   2    8     2     5   
26    a763dc6cd6db21bd301b7dd56d13d5019a6fd95fa5ff38...   1    7     4     2   
27    1b059745900e932f1393b71260faa4de9020b52c693860...   1    7     5     3   
28    3b24afa5378a11134ff6b8f1d2fc4a0313e0c1e6090bc6...   1    9     4     6   
29    4932d54890232be3d59b064a348e9dc510044d5526b990...   2    7     3     5   
...                                                 ...  ..  ...   ...   ...   
9970  5cf8917cbc9748f1864a2cc3194d333b650aa82948bcaa...   1    2     0     2   
9971  1106cab091690e13fecbc4a07c457142a8a85148e43dff...   1    1    99     1   
9972  7d6483cf84d466a3e9e046c1be79891a154a3811f815ee...   1    9    99     3   
9973  468911be1070fdba69fc161b485050ba9291a193b4e2d3...   2    3    99     1   
9974  5679d5dff3e805aa5cfcc194e69c44ef832374e97ef154...   1    5     3     2   
9975  326a0eb05f70a4981fcb5a86de96c2bdb51451386b471e...   2    1    99     1   
9976  1114dadf3d2b6fc9d6e50eaf7a91d2207199f2e9e1731f...   1    7     3     4   
9977  68e6fe3b7d7008a4ee598448ec139d09e274d1a2b618d2...   2    4     3     1   
9978  846dc503f7d7ac1d77646af6480115119425a5e8feb79e...   2    8    99     2   
9979  1e982e4a25585df76a175fe3d7d8aef35b1ce46cd6b823...   1    3     4     1   
9980  aae67ac5ae3cafeb9904705cc970584731e31a4c24c56f...   1    3     5     1   
9981  27a11a3c1d1abaa11c881d4e3cdaa2fbfd36b1dbed9bb5...   2    8    99     1   
9982  55677760ca600020ec4fc197148c1425df5d5c53415046...   1    2     4     1   
9983  5919f1968b57e9789dd35f36c2ae607a76f8dbb1dcf85c...   2    7     4     1   
9984  01d85a4888646186cbd0201843b888c4bc9818698491f4...   2    6    99     2   
9985  3b1bc9dc4f9c7d2276f5315b95a0f2240f2986b7033d84...   1    8    99     2   
9986  d2810a2b92f2980b667487f31e9b738f17d30aaf7d5321...   1    3     4     1   
9987  efb1354b2eb11ee578fa2db83e1a05f1d4322b9b9052b0...   2    5    99     3   
9988  b2df4d87d04c0685e6301bd4b202533d8b76669401f1d

In [330]:
df_imputer_ver2.to_csv("cus_info_missingvalue_out.csv", index= False)

### Baseline 모델로 모델 돌리기 

In [331]:
import pandas as pd
pd.set_option("display.max_row", 100)
pd.set_option("display.max_column", 100)
import numpy as np
import os

from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import mean_squared_error as mse
from sklearn.model_selection import train_test_split, StratifiedKFold, KFold
from lightgbm import LGBMRegressor

In [351]:
cus = pd.read_csv("/data/cus_info_missingvalue_out.csv")
iem = pd.read_csv("/data/iem_info_missingvalue_out.csv")
hist = pd.read_csv("/data/stk_bnc_hist.csv")
train = pd.read_csv("/data/stk_hld_train.csv")
test = pd.read_csv("/data/stk_hld_test.csv")

submission = pd.read_csv("/data/sample_submission.csv")

In [352]:
cus = cus.rename({'계좌ID':'act_id','성별':'sex_dit_cd','연령대':'cus_age_stn_cd',
                '투자성향':'ivs_icn_cd', '자산구간':'cus_aet_stn_cd', 
               '주거래상품군':'mrz_pdt_tp_sgm_cd','lifestage':'lsg_sgm_cd',
               '서비스등급':'tco_cus_grd_cd','총투자기간':'tot_ivs_te_sgm_cd',
              '주거래업종구분':'mrz_btp_dit_cd'}, axis = 1)

In [353]:
iem = iem.rename({'종목코드':'iem_cd','종목한글명':'iem_krl_nm', '종목업종': 'btp_cfc_cd',
              '시가총액규모유형':'mkt_pr_tal_scl_tp_cd','시장구분':'stk_dit_cd'}, axis=1)

In [354]:
iem.columns

Index(['iem_cd', 'iem_krl_nm', 'btp_cfc_cd', 'mkt_pr_tal_scl_tp_cd'], dtype='object')

In [357]:
iem

iem_cd             iem_krl_nm  btp_cfc_cd  mkt_pr_tal_scl_tp_cd
0     A000020                   동화약품           8                     2
1     A000080                  하이트진로          14                     2
2     A000180                 성창기업지주           5                     3
3     A000227                유유제약2우B           8                    99
4     A000325                 노루홀딩스우           2                     5
5     A000440                 중앙에너비스           7                     3
6     A000590                  CS홀딩스           2                     3
7     A000670                     영풍          11                     2
8     A000700                  유수홀딩스           5                     3
9     A001000                   신라섬유           7                     3
10    A001020                 페이퍼코리아          10                     3
11    A001070                   대한방직          14                     3
12    A001270                   부국증권           2                     3
13    A001290                  상상인증권           2                     3
14    A001390                  KG케미칼          12                     3
15    A001450                   현대해상           2                     2
16    A001460                    BYC          14                     3
17    A001465                   BYC우          14                     5
18    A001470                   삼부토건           1                     2
19    A001525                    동양우          11                     5
20    A001780                    알루코          11                     3
21    A001790                   대한제당          14                     3
22    A001799                대한제당3우B          14                    99
23    A001800                 오리온홀딩스           2                     2
24    A001810                   무림SP          10                     3
25    A002030                    아세아           5                     3
26    A002070                    비비안          14                     3
27    A002170                   삼양통상          14                     3
28    A002230                   피에스텍          10                     3
29    A002290                 삼일기업공사           1                     3
30    A002300                   한국제지          14                     3
31    A002310                  아세아제지          10                     3
32    A002360                SH에너지화학          12                     3
33    A002420                   세기상사           5                     3
34    A002460                   화성산업           1                     3
35    A002550                 KB손해보험          14                     2
36    A002710                  TCC스틸          11                     3
37    A002790                   아모레G          12                     1
38    A002840                   미원상사          12                     2
39    A002900                    TYM           3                     3
40    A002990                   금호건설           1                     3
41    A003010                     혜인           7                     3
42    A003075                코오롱글로벌우           1                     5
43    A003220                   대원제약           8                     3
44    A003480               한진중공업홀딩스           9                     3
45    A003530                 한화투자증권           2                     2
46    A003535                한화투자증권우           2                     5
47    A003547                대신증권2우B           2                    99
48    A003560                    IHQ           5                     3
49    A003580                넥스트사이언스          14                     3
...       ...                    ...         ...                   ...
3029  A340120                하이제5호스팩           2                     3
3030  A340360                   다보링크           2                     3
3031  A341790                 대한바이오팜          14                    99
3032  A341850    TIGER KIS부동산인프라채권TR          14                     4
3033  A342550              케이비제

In [355]:
cus.columns

Index(['act_id', 'sex_dit_cd', 'cus_age_stn_cd', 'ivs_icn_cd',
       'cus_aet_stn_cd', 'mrz_pdt_tp_sgm_cd', 'lsg_sgm_cd', 'tco_cus_grd_cd',
       'tot_ivs_te_sgm_cd', 'mrz_btp_dit_cd'],
      dtype='object')

In [356]:
test

act_id   iem_cd    byn_dt  \
0      0ad104dbed99be0cd858aa772765ddedade554601a981b...  A032640  20200522   
1      0ad104dbed99be0cd858aa772765ddedade554601a981b...  A160600  20190823   
2      0ad104dbed99be0cd858aa772765ddedade554601a981b...  A234340  20200611   
3      0ad104dbed99be0cd858aa772765ddedade554601a981b...  A131760  20200120   
4      0ad104dbed99be0cd858aa772765ddedade554601a981b...  A293490  20201217   
5      b6623032ed365b6c9a71c0344ebb8ca191f1ebc5927771...  A050110  20200814   
6      b6623032ed365b6c9a71c0344ebb8ca191f1ebc5927771...  A032640  20200805   
7      b6623032ed365b6c9a71c0344ebb8ca191f1ebc5927771...  A033320  20180126   
8      b6623032ed365b6c9a71c0344ebb8ca191f1ebc5927771...  A003570  20200902   
9      b6623032ed365b6c9a71c0344ebb8ca191f1ebc5927771...  A009580  20181206   
10     356b35c423a191e1f5b8cc1dfa2d597c614ba00b52ccae...  A128820  20200707   
11     64aae8dd71e5c0761000db9f9a6779e504e7a4aa9dc097...  A000030  20160104   
12     64aae8dd71e5c0761000db9f9a6779e504e7a4aa9dc097...  A316140  20190213   
13     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A036200  20201222   
14     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A005385  20201215   
15     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A013120  20201228   
16     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A152330  20200720   
17     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A181710  20200917   
18     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A005935  20191114   
19     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A344820  20201218   
20     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A120115  20201222   
21     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A222390  20160104   
22     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A028150  20190423   
23     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A000210  20201208   
24     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A000720  20201229   
25     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A001680  20200922   
26     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A017810  20200922   
27     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A303030  20200120   
28     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A043910  20190214   
29     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A033780  20191114   
30     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A014280  20200918   
31     5f7c3a8f37d9c268d06130ff0be5d32a1b9ef68c13049f...  A023410  20200918   
32     1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...  A005490  20170706   
33     1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...  A033780  20200623   
34     1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...  A032640  20201214   
35     1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...  A035420  20200908   
36     1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...  A005930  20200623   
37     1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...  A005940  20160104   
38     1119c23c3a504ca7b75060277410c0f6fb9018ec7638c2...  A008770  20200623   
39     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A011930  20160919   
40     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A103590  20160104   
41     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A069500  20190401   
42     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A101530  20160525   
43     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A114090  20160104   
44     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A041460  20160715   
45     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A003380  20180716   
46     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A083310  20160804   
47     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A100130  20160104   
48     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A200130  20160104   
49     6d497facfa1ea5901b827335553331f8555fec02a8184f...  A122350  20160104   
...                  

In [358]:
# 앞서 언급한 것처럼 베이스라인에서는 "hist_d" column을 임의로 생성하여 모델을 학습시키도록 하겠습니다.
# 베이스라인에서는 "hold_d"값, 즉 주식 보유기간의 0.6배에 해당하는 기간을 임의로 설정하여 "hist_d"를 생성하였습니다.
# 결국 모델은 "hist_d"만큼 주식을 보유 했을때의 "hold_d"를 예측하게 될 것입니다.

train["hist_d"] = train["hold_d"]*0.6
train.hist_d = np.trunc(train["hist_d"])

In [359]:
train.head(3)

act_id   iem_cd    byn_dt  \
0  0ad104dbed99be0cd858aa772765ddedade554601a981b...  A006360  20180726   
1  0ad104dbed99be0cd858aa772765ddedade554601a981b...  A005930  20180131   
2  0ad104dbed99be0cd858aa772765ddedade554601a981b...  A005070  20180517   

   hold_d  hist_d  
0      11     6.0  
1      80    48.0  
2       5     3.0

In [360]:
# train과 test에 고객정보(cus_info)와 주식정보(iem_info)를 추가하겠습니다.

train_data = pd.merge(train, cus, how = "left", on = ["act_id"])
train_data = pd.merge(train_data, iem, how = "left", on = ["iem_cd"])

test_data = pd.merge(test, cus, how = "left", on = ["act_id"])
test_data = pd.merge(test_data, iem, how = "left", on = ["iem_cd"])

In [361]:
train_data.head(3)

act_id   iem_cd    byn_dt  \
0  0ad104dbed99be0cd858aa772765ddedade554601a981b...  A006360  20180726   
1  0ad104dbed99be0cd858aa772765ddedade554601a981b...  A005930  20180131   
2  0ad104dbed99be0cd858aa772765ddedade554601a981b...  A005070  20180517   

   hold_d  hist_d  sex_dit_cd  cus_age_stn_cd  ivs_icn_cd  cus_aet_stn_cd  \
0      11     6.0           1               9           3               2   
1      80    48.0           1               9           3               2   
2       5     3.0           1               9           3               2   

   mrz_pdt_tp_sgm_cd  lsg_sgm_cd  tco_cus_grd_cd  tot_ivs_te_sgm_cd  \
0                  2           9               5                  5   
1                  2           9               5                  5   
2                  2           9               5                  5   

   mrz_btp_dit_cd iem_krl_nm  btp_cfc_cd  mkt_pr_tal_scl_tp_cd  
0               8       GS건설           1                     1  
1               8       삼성전자           9                     1  
2               8     코스모신소재          12                     2

In [362]:
# train_data에서 Y값을 추출한 후 hold_d column을 지워주겠습니다.

train_label = train_data["hold_d"]
train_data.drop(["hold_d"], axis = 1, inplace = True)

In [363]:
# 추가적으로 약간의 전처리를 통해 train data와 test data를 구성하겠습니다.

hist["stk_p"] = hist["tot_aet_amt"] / hist["bnc_qty"]
hist = hist.fillna(0)

train_data = pd.merge(train_data, hist, how = "left", on = ["act_id", "iem_cd"])
train_data = train_data[(train_data["byn_dt"] == train_data["bse_dt"])]
train_data.reset_index(drop = True, inplace = True)

test_data = pd.merge(test_data, hist, how = "left", on = ["act_id", "iem_cd"])
test_data = test_data[(test_data["byn_dt"] == test_data["bse_dt"])]
test_data.reset_index(drop = True, inplace = True)

train_data = train_data.drop(["act_id", "iem_cd", "byn_dt", "bse_dt"], axis = 1)
test_data = test_data.drop(["act_id", "iem_cd", "byn_dt", "submit_id", "hold_d", "bse_dt"], axis = 1)

L_encoder = LabelEncoder()
L_encoder.fit(iem["iem_krl_nm"])
train_data["iem_krl_nm"] = L_encoder.transform(train_data["iem_krl_nm"])
test_data["iem_krl_nm"] = L_encoder.transform(test_data["iem_krl_nm"])

In [364]:
train_data.head(3)

hist_d  sex_dit_cd  cus_age_stn_cd  ivs_icn_cd  cus_aet_stn_cd  \
0     6.0           1               9           3               2   
1    48.0           1               9           3               2   
2     3.0           1               9           3               2   

   mrz_pdt_tp_sgm_cd  lsg_sgm_cd  tco_cus_grd_cd  tot_ivs_te_sgm_cd  \
0                  2           9               5                  5   
1                  2           9               5                  5   
2                  2           9               5                  5   

   mrz_btp_dit_cd  iem_krl_nm  btp_cfc_cd  mkt_pr_tal_scl_tp_cd  bnc_qty  \
0               8         101           1                     1    274.0   
1               8        1361           9                     1      2.0   
2               8        2530          12                     2    786.0   

   tot_aet_amt  stk_par_pr      stk_p  
0   11782000.0      5000.0    43000.0  
1    4990000.0      5000.0  2495000.0  
2   14619600.0      1000.0    18600.0

In [365]:
test_data.head(3)

hist_d  sex_dit_cd  cus_age_stn_cd  ivs_icn_cd  cus_aet_stn_cd  \
0     153           1               9           3               2   
1     335           1               9           3               2   
2     139           1               9           3               2   

   mrz_pdt_tp_sgm_cd  lsg_sgm_cd  tco_cus_grd_cd  tot_ivs_te_sgm_cd  \
0                  2           9               5                  5   
1                  2           9               5                  5   
2                  2           9               5                  5   

   mrz_btp_dit_cd  iem_krl_nm  btp_cfc_cd  mkt_pr_tal_scl_tp_cd  bnc_qty  \
0               8         418           4                     1    300.0   
1               8        2230          10                     3    198.0   
2               8        1515          13                     2    138.0   

   tot_aet_amt  stk_par_pr    stk_p  
0    3945000.0      5000.0  13150.0  
1    2524500.0       500.0  12750.0  
2    4291800.0       500.0  31100.0

In [366]:
train_data.reset_index(drop = True, inplace=True)
train_label.reset_index(drop = True, inplace=True)

In [372]:
models = []

folds = KFold(n_splits=10)
for train_idx, val_idx in folds.split(train_data):
    
    train_x = train_data.iloc[train_idx, :]
    train_y = train_label[train_idx]
    val_x = train_data.iloc[val_idx, :]
    val_y = train_label[val_idx]
    
    model = LGBMRegressor(objective= "regression",
                          max_depth= 8,
                          n_estimators= 2000,
                          learning_rate= 0.01,
                          num_leaves = 31)
    
    model.fit(train_x, train_y,
              eval_set=[(val_x, val_y)],
              eval_metric=["rmse"],
              early_stopping_rounds=300,
              verbose=500)
    
    models.append(model)

Training until validation scores don't improve for 300 rounds
[500]	valid_0's rmse: 4.21693	valid_0's l2: 17.7825
[1000]	valid_0's rmse: 4.08441	valid_0's l2: 16.6824
[1500]	valid_0's rmse: 4.04913	valid_0's l2: 16.3955
[2000]	valid_0's rmse: 4.03698	valid_0's l2: 16.2972
Did not meet early stopping. Best iteration is:
[2000]	valid_0's rmse: 4.03698	valid_0's l2: 16.2972
Training until validation scores don't improve for 300 rounds
[500]	valid_0's rmse: 1.2419	valid_0's l2: 1.54233
Early stopping, best iteration is:
[698]	valid_0's rmse: 1.17745	valid_0's l2: 1.38638
Training until validation scores don't improve for 300 rounds
[500]	valid_0's rmse: 0.685671	valid_0's l2: 0.470145
[1000]	valid_0's rmse: 0.477619	valid_0's l2: 0.22812
Early stopping, best iteration is:
[794]	valid_0's rmse: 0.47717	valid_0's l2: 0.227692
Training until validation scores don't improve for 300 rounds
[500]	valid_0's rmse: 0.574361	valid_0's l2: 0.32989
[1000]	valid_0's rmse: 0.413046	valid_0's l2: 0.17060

In [373]:
result = []
for i in models:
    result.append(i.predict(test_data))
predict = np.mean(result, axis = 0)

In [376]:
predict[:20]

array([254.68040083, 557.48727709, 232.15693195, 394.04688436,
        15.4831043 , 155.42101873, 167.33719553, 903.43599226,
       135.42695092, 874.78361919, 202.29902623, 890.17526114,
       894.3838571 ,  10.47837206,  19.00248032,   5.46198369,
       187.35475245, 117.45315004, 467.21193074,  14.0093925 ])

In [377]:
submission["hold_d"] = np.round(predict)

In [378]:
submission.to_csv("dacon_baseline_missinvalue_ver2.csv", index = False)